rfcs/0012-compaction-state-persistence.md

```diff
@@ -0,0 +1,487 @@
+6. When a task completes compaction execution, the task returns the {destinationId, outputSSTs} to the `CompactionEventHandler`
+7. `CompactionEventHandler` is responsible to update in-memory compaction state, cleanup the job task, write logs and update manifest[Need to update the compaction persistence state]
+8. GC clears the orphaned states and SSTs during it's run.
+
+### **Key Design Decisions**
+
+#### **1. Persistence Boundaries**
+**Decision**: Persist state at the critical boundary:
+- **Output SST Completion**: Every ~256MB of written data (always persisted)
+
+**Rationale**: Output SST completions provide the best recovery value per persistence operation. Each represents significant completed work that we don't want to lose.
+
+#### **2. Enhanced Job Model**  
+**Decision**: Change from 1:1 to 1:N relationship between Compaction and CompactionJob.
+
+**Rationale**: Enables retry logic, progress tracking, and recovery without breaking existing compaction scheduling logic.
+
+#### **3. State Management Pattern**
+**Decision**: Mirror the existing `ManifestStore` pattern with `CompactorStore`.
+
+**Rationale**: Reuses proven patterns for atomic updates, version checking, and conflict resolution that are already battle-tested in SlateDB.
+
+#### **4. Recovery Strategy**
+**Decision**: Smart recovery based on work preservation efficiency:
+- **High efficiency** (>80% work preserved): Resume from last completed output SST
+- **Low efficiency**: Full retry with exponential backoff
+
+**Rationale**: Balances complexity vs. value - don't over-optimize recovery for cases where restart is simpler.
+
+### **Data Model**
+```
+Compaction (1) ──→ (N) CompactionJob
+    ├── sources: Vec<SourceId>
+    ├── destination: u32  
+    ├── job_attempts: Vec<CompactionJob>
+    └── current_phase: CompactionPhase
+
+CompactionJob
+    ├── attempt_number: u32
+    ├── progress: CompactionProgress  
+    ├── status: CompactionJobStatus
+    └── existing_output_ssts: Option<Vec<SsTableId>>
+
+CompactionProgress
+    ├── input_ssts_completed: Vec<SsTableId>
+    ├── output_ssts_written: Vec<SsTableId>
+    ├── bytes_read/written: u64
+    └── current_phase: CompactionPhase
+```
+
+### **Persistent State Storage**
+
+#### **Object Store Layout**
+The compaction state is persisted to the object store following the same CAS pattern as manifests, ensuring consistency and reliability:
+
+```
+/compactor_state_000000001.fbs  # First compactor state
+/compactor_state_000000002.fbs  # Updated state after compactions
+/compactor_state_000000003.fbs  # Current state
+```
+
+#### **CompactionState Structure**
+The persistent state contains the complete view of all compaction activity:
```

Comment:
It would be helpful to have the .fbs definitions as well. This will help me understand the relationship between in-memory and in-object store state.

We normally have `.fbs` files, which get generated into `.rs` files with `flatc`. Then we have a codec (`flatbuffer_types.rs`), which translates the raw FlatBuffer-generated Rust structs into our own easy-to-use, in-memory structs. For example, manifest.fbs (ManifestV1) ➜ manifest_generated.rs (ManifestV1) ➜ flatbuffer_types.rs (FlatBufferManifestCodec) ➜ mod.rs (Manifest).